teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -84,97 +84,399 @@ that can be shared between them, and running the second Task will be delayed by
+    [a pod's containers cannot be updated](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core)
+  * Running each Task with a different `ServiceAccount` - the pod has one ServiceAccount as a whole
+
+(See also [functionality supported by experimental Pipeline to TaskRun](https://github.com/tektoncd/experimental/tree/main/pipeline-to-taskrun#supported-pipeline-features))
+
+##### Pipeline executed as pod
+
+In this implementation, when a Pipeline is executed as a pod, the Tekton Pipelines controller would construct a pod
+that implements that Pipeline.
+
+Pros:
+* Uses [an existing abstraction](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability)
+  (Pipelines)
+* Pipelines already have syntax for expressing some of the features we'd likely want for this functionality, e.g.
+  `finally`
+Cons:
+* Using an existing abstraction (Pipelines) could be confusing if we can't support all of a Pipeline's functionality
+  when running as a pod (which is likely)
+
+##### Pipeline executed as TaskRun
+
+This is the approach currently taken in the
+[Pipeline to TaskRun experimental custom task](https://github.com/tektoncd/experimental/tree/main/pipeline-to-taskrun).
+
+Cons:
+* We would be [limited in the features we could support](https://github.com/tektoncd/experimental/tree/main/pipeline-to-taskrun#supported-pipeline-features)
+  to features that TaskRuns already support, or we'd have to add more Pipeline features to TaskRuns.
+
+##### TaskRun controller composes one pod from multiple TaskRuns
+
+The goal behind this implementation would be that we still create multiple TaskRuns for a Pipeline, but those TaskRuns
+ultimately get executed as one pod.
+
+A pod will start executing once it has been created and many of the fields (including the containers list)
+[cannot be updated](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core).
+If we wanted to use one pod to execute an entire Pipeline, we would need to be aware of all of the containers that need
+to be in that pod before the pod was created. The PipelineRun controller currently only creates TaskRuns which are
+ready to executed, but in order to the pod which represents all of the TaskRuns that will ultimately represent the
+PipelineRun, the Pipeline controller would need to create all TaskRuns at once, with the TaskRuns that are not ready
+to execute in a pending state.
+
+As the TaskRun controller (or some new controller for this specific purpose) reconciled the TaskRuns that make up the
+Pipeline, it would need to:
+1. Be able to save state across reconcile loops that would represent the eventual pod and update it as it reconciled.
+   Since the relevant fields in the pod itself [cannot be updated](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core)
+   this would have to be done somewhere else, e.g. in memory in the controller or in `ConfigMap`.
```

Comment:
If we had a group CRD and group controller, the group controller would analyse the group when its first reconciled, and it would create a `Pod` with one container for each `TaskRun` that might be executed as part of the group.
It would then start creating `TaskRuns` to share the current state towards both users. It would add a link to created `TaskRuns` in its own status to trigger reconcile on `PipelineRun` controller and corresponding updates of the `PipelineRun`. Task results (within the group) could be stored on the `TaskRuns` as usual, or with the mechanism we come up with in the results TEP. 
Any other group orchestration specific data that is needed by the `Pod` could be stored on the `TaskGroup` CRD itself - I don't have an example in mind though of what kind of data this could be.